Serializer problem in win

Serializer crashes with:

```Bash
C:\KratosSource\kratos\includes/serializer.h(207): error C2039: 'load': is not a member of 'std::_Array_iterator<_Ty,1>

with
          [
              _Ty=Kratos::Table<double,double,1>::result_type
          ]
```

Seems to be related with Kratos::Tables. @pooyan-dadvand any idea?

--- a/kratos/includes/serializer.h
+++ b/kratos/includes/serializer.h
@@ -320,6 +320,14 @@ public:
         pVariable = static_cast<const Variable<TDataType>*>(GetVariableData(name));
     }
 
+	template<class TDataType, std::size_t TDataSize>
+	void load(std::string const & rTag, std::array<TDataType, TDataSize>& rObject)
+	{
+		load_trace_point(rTag);
+		for (SizeType i = 0; i < TDataSize; i++)
+			load("E", rObject[i]);
+	}
+
     template<class TDataType>
     void load(std::string const & rTag, std::vector<TDataType>& rObject)
     {
@@ -432,7 +440,13 @@ public:
     KRATOS_SERIALIZATION_DIRECT_LOAD(std::size_t)
 #endif
 
-
+	template<class TDataType, std::size_t TDataSize>
+	void save(std::string const & rTag, std::array<TDataType, TDataSize> const& rObject)
+	{
+		save_trace_point(rTag);
+		for (SizeType i = 0; i < TDataSize; i++)
+			save("E", rObject[i]);
+	}
 
     template<class TDataType>
     void save(std::string const & rTag, std::vector<TDataType> const& rObject)

--- a/kratos/includes/serializer.h
+++ b/kratos/includes/serializer.h
@@ -2,13 +2,13 @@
 //    ' /   __| _` | __|  _ \   __|
 //    . \  |   (   | |   (   |\__ `
 //   _|\_\_|  \__,_|\__|\___/ ____/
-//                   Multi-Physics 
+//                   Multi-Physics
 //
-//  License:		 BSD License 
+//  License:		 BSD License
 //					 Kratos default license: kratos/license.txt
 //
 //  Main authors:    Pooyan Dadvand
-//                    
+//
 //
 
 #if !defined(KRATOS_SERIALIZER_H_INCLUDED )
@@ -112,9 +112,9 @@ template <class TDataType> class Variable;
  *
  * @brief The serialization consists in storing the state of an object into a storage format like data file or memory buffer and also retrieving the object from such a media.
  *
- * @details The serialization consists in storing the state of an object into a storage format like data file or memory buffer and also retrieving the object from such a media. 
- * The idea of serialization is based on saving all object's data consecutively in the file or buffer and then load it in the same order. 
- * In Kratos a serialization mechanism is used for creating the restart file. So for storing an object into restart file and retrieve it afterward on must add the necessary component used by serialization. 
+ * @details The serialization consists in storing the state of an object into a storage format like data file or memory buffer and also retrieving the object from such a media.
+ * The idea of serialization is based on saving all object's data consecutively in the file or buffer and then load it in the same order.
+ * In Kratos a serialization mechanism is used for creating the restart file. So for storing an object into restart file and retrieve it afterward on must add the necessary component used by serialization.
  *
  * @author Pooyan Dadvand
  */
@@ -166,7 +166,7 @@ public:
             p_file = new std::fstream(std::string(Filename+".rest").c_str(), std::ios::binary|std::ios::out);
         }
         mpBuffer = p_file;
-        KRATOS_ERROR_IF_NOT(*mpBuffer) << "Error opening input file : " 
+        KRATOS_ERROR_IF_NOT(*mpBuffer) << "Error opening input file : "
                                        << std::string(Filename+".rest") << std::endl;
     }
 
@@ -233,7 +233,7 @@ public:
                     typename RegisteredObjectsContainerType::iterator i_prototype =  msRegisteredObjects.find(object_name);
 
                     KRATOS_ERROR_IF(i_prototype == msRegisteredObjects.end())
-                        << "There is no object registered in Kratos with name : " 
+                        << "There is no object registered in Kratos with name : "
                         << object_name << std::endl;
 
                     if(!pValue)
@@ -394,13 +394,8 @@ public:
     void load(std::string const & rTag, bounded_vector<TDataType, TDimension>& rObject)
     {
         load_trace_point(rTag);
-        SizeType size;
 
-        load("size", size);
-
-        KRATOS_DEBUG_ERROR_IF_NOT(rObject.size() == size) << "Object has wrong size!" << std::endl;
-
-        for(SizeType i = 0 ; i < size ; ++i)
+        for(SizeType i = 0 ; i < TDimension ; ++i)
             load("E", rObject[i]);
 //    read(rObject);
     }
@@ -409,17 +404,9 @@ public:
     void load(std::string const & rTag, bounded_matrix<TDataType, TDimension1, TDimension2>& rObject)
     {
         load_trace_point(rTag);
-        SizeType size1;
-        SizeType size2;
 
-        load("size1", size1);
-        load("size2", size2);
-
-        KRATOS_DEBUG_ERROR_IF_NOT(rObject.size1() == size1 && 
-                                  rObject.size2() == size2) << "Object has wrong size!" << std::endl;
-        
-        for(SizeType i = 0 ; i < size1 ; ++i)
-            for(SizeType j = 0 ; j < size2 ; ++j)
+        for(SizeType i = 0 ; i < TDimension1 ; ++i)
+            for(SizeType j = 0 ; j < TDimension2 ; ++j)
                 load("E", rObject(i,j));
 //    read(rObject);
     }
@@ -618,11 +605,8 @@ public:
     void save(std::string const & rTag, bounded_vector<TDataType, TDimension> const& rObject)
     {
         save_trace_point(rTag);
-        SizeType size = rObject.size();
 
-        save("size", size);
-
-        for(SizeType i = 0 ; i < size ; ++i)
+        for(SizeType i = 0 ; i < TDimension ; ++i)
             save("E", rObject[i]);
 //    write(rObject);
     }
@@ -631,14 +615,9 @@ public:
     void save(std::string const & rTag, bounded_matrix<TDataType, TDimension1, TDimension2> const& rObject)
     {
         save_trace_point(rTag);
-        SizeType size1 = rObject.size1();
-        SizeType size2 = rObject.size2();
 
-        save("size1", size1);
-        save("size2", size2);
-
-        for(SizeType i = 0 ; i < size1 ; ++i)
-            for(SizeType j = 0 ; j < size2 ; ++j)
+        for(SizeType i = 0 ; i < TDimension1 ; ++i)
+            for(SizeType j = 0 ; j < TDimension2 ; ++j)
                 save("E", rObject(i,j));
 //    write(rObject);
     }
@@ -897,7 +876,7 @@ private:
                 typename RegisteredObjectsNameContainerType::iterator i_name = msRegisteredObjectsName.find(typeid (*pValue).name());
 
                 if (i_name == msRegisteredObjectsName.end())
-                    KRATOS_ERROR << "There is no object registered in Kratos with type id : " 
+                    KRATOS_ERROR << "There is no object registered in Kratos with type id : "
                                  << typeid (*pValue).name() << std::endl;
                 else
                     write(i_name->second);

--- a/kratos/includes/table.h
+++ b/kratos/includes/table.h
@@ -785,7 +785,7 @@ private:
         for(auto i_row = mData.begin() ; i_row != mData.end() ; i_row++){
             rSerializer.save("Argument", i_row->first);
             for(auto j = i_row->second.begin() ; j != i_row->second.end(); j++)
-                rSerializer.save("Column", j);
+                rSerializer.save("Column", *j);
         }
     }
 
@@ -800,7 +800,7 @@ private:
         for(auto i_row = mData.begin() ; i_row != mData.end() ; i_row++){
             rSerializer.load("Argument", i_row->first);
             for(auto j = i_row->second.begin() ; j != i_row->second.end() ; j++)
-                rSerializer.load("Column", j);
+                rSerializer.load("Column", *j);
         }
    }
 

--- a/kratos/tests/includes/test_serializer.cpp
+++ b/kratos/tests/includes/test_serializer.cpp
@@ -45,8 +45,7 @@ namespace Kratos {
         {
             SaveAndLoadObjects(rObjectToBeSaved, rObjectToBeLoaded);
 
-            KRATOS_ERROR_IF_NOT(rObjectToBeSaved.size() == rObjectToBeLoaded.size()) 
-                << "Object sizes are not consistent" << std::endl;
+            KRATOS_CHECK_EQUAL(rObjectToBeLoaded.size(), rObjectToBeSaved.size());
 
             for (std::size_t i=0; i< rObjectToBeSaved.size(); ++i)
                 KRATOS_CHECK_EQUAL(rObjectToBeLoaded[i], rObjectToBeSaved[i]);
@@ -55,11 +54,10 @@ namespace Kratos {
         template<typename TObjectType>
         void TestObjectSerializationComponentwise2D(const TObjectType& rObjectToBeSaved, TObjectType& rObjectToBeLoaded)
         {
-            SaveAndLoadObjects(rObjectToBeSaved, rObjectToBeLoaded);   
+            SaveAndLoadObjects(rObjectToBeSaved, rObjectToBeLoaded);
 
-            KRATOS_ERROR_IF_NOT(rObjectToBeSaved.size1() == rObjectToBeLoaded.size1() &&
-                                rObjectToBeSaved.size2() == rObjectToBeLoaded.size2()) 
-                << "Object sizes are not consistent" << std::endl;
+            KRATOS_CHECK_EQUAL(rObjectToBeLoaded.size1(), rObjectToBeSaved.size1());
+            KRATOS_CHECK_EQUAL(rObjectToBeLoaded.size2(), rObjectToBeSaved.size2());
 
             for (std::size_t i=0; i<rObjectToBeSaved.size1(); ++i) {
                 for (std::size_t j=0; j<rObjectToBeSaved.size2(); ++j) {
@@ -86,14 +84,14 @@ namespace Kratos {
         }
 
         /*********************************************************************/
-        /* Testing the Datatypes that for which the 
+        /* Testing the Datatypes that for which the
            "KRATOS_SERIALIZATION_DIRECT_LOAD" macro is used */
         /*********************************************************************/
         KRATOS_TEST_CASE_IN_SUITE(SerializerBool, KratosCoreFastSuite)
         {
             bool object_to_be_saved = true;
             bool object_to_be_loaded;
-            
+
             TestObjectSerialization(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -101,7 +99,7 @@ namespace Kratos {
         {
             int object_to_be_saved = -105;
             int object_to_be_loaded;
-            
+
             TestObjectSerialization(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -109,7 +107,7 @@ namespace Kratos {
         {
             long object_to_be_saved = -1598456605;
             long object_to_be_loaded;
-            
+
             TestObjectSerialization(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -117,7 +115,7 @@ namespace Kratos {
         {
             double object_to_be_saved = -159845.6605;
             double object_to_be_loaded;
-            
+
             TestObjectSerialization(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -125,7 +123,7 @@ namespace Kratos {
         {
             unsigned long object_to_be_saved = 1598456605;
             unsigned long object_to_be_loaded;
-            
+
             TestObjectSerialization(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -133,7 +131,7 @@ namespace Kratos {
         {
             unsigned int object_to_be_saved = 15905;
             unsigned int object_to_be_loaded;
-            
+
             TestObjectSerialization(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -141,7 +139,7 @@ namespace Kratos {
         {
             std::string object_to_be_saved = "MyStringToBeSerialized";
             std::string object_to_be_loaded;
-            
+
             TestObjectSerialization(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -151,7 +149,7 @@ namespace Kratos {
             Matrix object_to_be_loaded;
 
             FillMatrixWithValues(object_to_be_saved);
-            
+
             TestObjectSerializationComponentwise2D(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -159,7 +157,7 @@ namespace Kratos {
         {
             long long object_to_be_saved = -1598456546843565605;
             long long object_to_be_loaded;
-            
+
             TestObjectSerialization(object_to_be_saved, object_to_be_loaded);
         }
 
@@ -191,6 +189,23 @@ namespace Kratos {
                 KRATOS_CHECK_EQUAL((*p_loaded_array)[i], (*p_array)[i]);
         }
 
+        KRATOS_TEST_CASE_IN_SUITE(SerializerStdArray, KratosCoreFastSuite)
+        {
+            using Array5Dbl = std::array<double,5>;
+            using Array7Int = std::array<int,7>;
+
+            Array5Dbl object_to_be_saved_1;
+            Array5Dbl object_to_be_loaded_1;
+            Array7Int object_to_be_saved_2;
+            Array7Int object_to_be_loaded_2;
+
+            FillVectorWithValues(object_to_be_saved_1);
+            FillVectorWithValues(object_to_be_saved_2);
+
+            TestObjectSerializationComponentwise1D(object_to_be_saved_1, object_to_be_loaded_1);
+            TestObjectSerializationComponentwise1D(object_to_be_saved_2, object_to_be_loaded_2);
+        }
+
         KRATOS_TEST_CASE_IN_SUITE(SerializerStdVector, KratosCoreFastSuite)
         {
             std::vector<double> object_to_be_saved(5);
@@ -213,9 +228,9 @@ namespace Kratos {
 
         KRATOS_TEST_CASE_IN_SUITE(SerializerMap, KratosCoreFastSuite)
         {
-            std::map <std::string, double> object_to_be_saved { 
-                { "42", -30.556 }, 
-                { "3",   10.258 } 
+            std::map <std::string, double> object_to_be_saved {
+                { "42", -30.556 },
+                { "3",   10.258 }
             };
             std::map<std::string, double> object_to_be_loaded;
 
@@ -224,9 +239,9 @@ namespace Kratos {
 
         KRATOS_TEST_CASE_IN_SUITE(SerializerUnorderedMap, KratosCoreFastSuite)
         {
-            std::unordered_map <std::string, double> object_to_be_saved { 
-                { "42", -30.556 }, 
-                { "3",   10.258 } 
+            std::unordered_map <std::string, double> object_to_be_saved {
+                { "42", -30.556 },
+                { "3",   10.258 }
             };
             std::unordered_map<std::string, double> object_to_be_loaded;
 
@@ -273,7 +288,7 @@ namespace Kratos {
         {
             using Matrix53Double = bounded_matrix<double,5,3>;
             using Matrix26Int = bounded_matrix<int,2,6>;
-            
+
             Matrix53Double object_to_be_saved_1;
             Matrix53Double object_to_be_loaded_1;
             Matrix53Double object_to_be_saved_2;
@@ -281,7 +296,7 @@ namespace Kratos {
 
             FillMatrixWithValues(object_to_be_saved_1);
             FillMatrixWithValues(object_to_be_saved_2);
-            
+
             TestObjectSerializationComponentwise2D(object_to_be_saved_1, object_to_be_loaded_1);
             TestObjectSerializationComponentwise2D(object_to_be_saved_2, object_to_be_loaded_2);
         }

